[docs] Add internal link rewrite MDX plugin

byCedric opened the pull request:
**Edit:** simplification is done, a stacked PR containing all url rewrite and fixes is pending (see #10983).

# Why

- We don't have tooling to validate if internal links still work, e.g. referencing proper file or header
- When we switched internal MD(X) links to `next/link`, we encountered some issues related to the difference between MD(X) non-trailing-slash and Next trailing-slash behavior. e.g. `./path/` and `../path/` weren't working anymore.

This PR fixes both. 😄 

Link validation | VSCode linking
--- | ---
![mdx-link-validation](https://user-images.githubusercontent.com/1203991/98680294-7cf28300-2361-11eb-86c6-17ccb1a3f54a.gif) | ![mdx-vscode-linking](https://user-images.githubusercontent.com/1203991/98680299-7e23b000-2361-11eb-9904-6fe197513537.gif)

# How

This link rewriter only modifies internal links between the pages, in MDX only. Anchor or HTML links aren't modified because it's only implemented as remark plugin (not rehype, html transformer).

It allows us to reference files instead of absolute urls, allowing us to implement tooling around validation of these links. ~~I implemented it to be backwards compatible with our current links, although the tooling can't find those files and will warn that links aren't valid.~~

With this, we can write links like:

### From `/push-notifications/overview.md`, to `/push-notifications/using-fcm.md`
- `using-fcm` → link validation only works with `.md` reference
- ~~`../using-fcm/` → also link validation failure, but our current way of linking~~
- `../using-fcm.md` → link validation failure, hard-codes trailing-slash behavior, file isn't located `../`
- `./using-fcm.md` → fully picked up by auto validator
- `using-fcm.md` → fully picked up by auto validator
- `using-fcm.md#some-header` → fully picked up by auto validator

# Test Plan

- See unit tests for the rewriter
- See commits as paper-trail for the changes
- `$ yarn lint-links` to validate internal links (takes ±30s)
- `$ yarn dev`
	1. check the index page
	2. check the push-notifications overview page, first 3 links in list

# ~~TODO~~

~~Before merging this, I'd love to simplify the setup by updating the internal linking to follow 1 format. Preferably one that's supported by the `remark -u validate-links` command. With that, we can simplify the rewriter by _a lot_.~~

~~It's also weird that paths like `../../workflow/linking` works from `/versions/v39.0.0/sdk/app-auth.md`. It should link to `/versions/workflow/linking`, but the error page probably redirects to the proper url (`/workflow/linking`). That's also something I'd like to see fixed.~~

byCedric committed:
[docs] Move custom remark heading plugin to mdx plugins

byCedric committed:
[docs] Add custom MDX to Next link rewriter

byCedric committed:
[docs] Add direct file references as example

byCedric committed:
[docs] Add new mdx remark plugin to next config

byCedric committed:
[docs] Add remark tool to validate internal links

byCedric committed:
[docs] Re-enable internal links in MDX

barthap commented on docs/mdx-plugins/remark-link-rewrite.test.js:
```diff
@@ -0,0 +1,103 @@
+describe('from index page', () => {
+  const file = makeFile('index.md');
+
+  it('resolves guides to guides/', () => {
+    expect(rewrite(file, 'guides')).toBe('guides');
+  });
```
The name of the test is confusing because of that trailing slash, which then doesn't appear in assertion.

barthap commented on docs/mdx-plugins/remark-link-rewrite.test.js:
```diff
@@ -0,0 +1,103 @@
+  });
+
+  // legacy method of linking, hardcoding trailing slash
+  it('resolves ../guide.md to /guides/', () => {
```
```suggestion
  it('resolves ../guides.md to /guides/', () => {
```

barthap commented on docs/mdx-plugins/remark-link-rewrite.test.js:
```diff
@@ -0,0 +1,103 @@
+  });
+
+  // legacy method of linking, hardcoding trailing slash
+  it('resolves ../guide/ to /guides/', () => {
```
```suggestion
  it('resolves ../guides/ to /guides/', () => {
```

barthap commented on docs/mdx-plugins/remark-link-rewrite.test.js:
```diff
@@ -0,0 +1,103 @@
+describe('from nested page', () => {
+  const file = makeFile('push-notifications/overview.md');
+
+  it('resolves using-fcm to using-fcm/', () => {
+    expect(rewrite(file, 'using-fcm')).toBe('using-fcm');
+  });
```
Same here

barthap reviewed (approved):
Just found a few typos in tests.

To make sure I undestand correctly:
- Until now we used internal relative paths containing `./` and `../` with trailing shash
- Now we can use relative file paths with `.md` suffix
- Remark validator `yarn lint-links` passes the `.md` links but will warn on the old links. It should be run _before_ running `yarn dev` or `yarn build`.
- The new remark plugin rewrites all possible cases into links valid for next-js. It replaces `./` and `../` with proper path, removes `.md` extension and adds a trailing slash.
- Those markdown changes are just an example how we should handle all links

byCedric commented:
> * Until now we used internal relative paths containing `./` and `../` with trailing shash
It doesn't need the trailing slash, Next auto-adds it if it's not there. We mostly use `../`, not `./`.

> * Now we can use relative file paths with `.md` suffix
Yes, direct file reference. So instead of hardcoding Next `trailingSlash`, we translate the actual file path to Next url with trailing slash. The main difference is that Next handles files as folders, while MD(X) handles files as files.

> * Remark validator `yarn lint-links` passes the `.md` links but will warn on the old links. It should be run _before_ running `yarn dev` or `yarn build`.
Yes, we can run it before or after. It's just a tool for us to validate if all links are correct 😄 

> * The new remark plugin rewrites all possible cases into links valid for next-js. It replaces `./` and `../` with proper path, removes `.md` extension and adds a trailing slash.
Yes:
- `.md` extensions are translated to `/` to convert it from MD(X) "file behavior" to Next "folder behavior"
- `../` is replaced with the actual file URL from `/` to make `next/link` work with these urls.

> * Those markdown changes are just an example how we should handle all links
Yes 😄

cruzach commented on docs/mdx-plugins/remark-link-rewrite.js:
```diff
@@ -0,0 +1,100 @@
+
+  return (tree, file) => {
+    // get the relative path, from pages dir to the current file
+    const filePrefix = getFilePrefix(settings, file);
```
`filePrefix` could be `undefined` here, right? would it be better to return an empty string in the case that there is no file prefix (i.e. if (fileUrl === ignored))

byCedric replied:
The rewriter now quits early when the file locations aren't defined. 😄

cruzach reviewed (changes requested):
It's probably a good idea to add a section on internal links to the [docs readme](https://github.com/expo/expo/blob/master/docs/README.md), that way we have a best practice in regards to links that people can start to follow (and eventually we can get rid of the code that's just here for backwards compatibility 🤞 )

brentvatne commented on docs/mdx-plugins/remark-link-rewrite.js:
```diff
@@ -0,0 +1,100 @@
+ * If the path starts with either `./` or `../` we need to clean this.
+ * Next doesn't work nicely with these kinds of paths.
+ *   - from: /push-notifications/overview/, to: sending-notifications/ -> sending-notifications/
+ *   - from: /push-notifications/overview/, to: ./sending-notifications/ -> /push-notifications/sending-notifications/
```
this particular case seems odd to me. it may end up to be confusing that `./sending-notifications/` and `../sending-notifications/` are both resolved to the same path `/push-notifications/sending-notifications/`. maybe we need to update all occurrences of one of these cases.

byCedric replied:
Yes, this is the confusing backwards compatibility with hardcoded trailing slash. I'll continue on this PR to remove that behavior, and add a stacked PR to change the current usage. The rewriter will basically be the same behavior Github has when linking to files directly, that makes it so much more understandable.

brentvatne reviewed (commented):
it seems sensible to move towards internal links pointing to md files directly, this simplifies the thought process of internal linking -- just provide a relative path to the file you want to link to. thanks for digging in to this!

byCedric committed:
[docs] Simplify internal MDX link rewrite

byCedric committed:
[docs] Add readme note about internal linking

byCedric committed:
[docs] Add link validation to workflow after export

We need to perform this after an export to make sure /versions/latest exists.

cruzach reviewed (approved):
looks good 👍 thanks for adding the section to the readme

Do we need to land this and https://github.com/expo/expo/pull/10983 together before deploying the docs?

byCedric commented:
> Do we need to land this and #10983 together before deploying the docs?

Yes, we have to land it together. So preferably, the stacked one first and then this one. Else we have the same issue related to internal links not working properly on e.g. push notifications guides. 

I'm working on that one, but I still have to fix a lot of "wrong file references". We can skip the header references, and fix that in a different PR. It's not good, but not that bad if some anchor links aren't working.

bbarthec commented on docs/mdx-plugins/remark-link-rewrite.js:
```diff
@@ -0,0 +1,64 @@
+ * @returns {function} remark plugin
+ */
+module.exports = function remarkLinkRewrite(options) {
+  const settings = Object.assign({}, DEFAULT_OPTIONS, options);
```
Hmm, why not to go with the `newer` approach? 🤔 
```suggestion
  const settings = { ...DEFAULT_OPTIONS, ...options };
```

byCedric replied:
You are right, I just copied this from an example `remark-` plugin.

byCedric committed:
[docs] Use object spread in remark link rewrite

Co-authored-by: Bartłomiej Bukowski <bartlomiej.bukowski@swmansion.com>

byCedric committed:
[docs] Ignore react-native internal linking

byCedric commented:
Alright, I'm going to merge the two PRs (this one, and the one that refactors the internal links). I've reverted the changes in `react-native` versioned folders because they _should_ already work as expected. We can always re-enable linting on those links as well, but for now I think it's best to limit the amount of changes.

Just rebased both on master and fixed the remaining stuff.

byCedric committed:
[docs] Rewrite all internal links to file references (#10983)

* [docs] Rewrite all internal links to file references

* [docs] Manually fix last remaining broken internal links

* [docs] Manually fix last remaining broken links to repo assets

* [docs] Fix last batch of internal links

* [docs] Fix final batch of internal links

* [docs] Revert link changes in react-native docs
